### PR TITLE
bwpackedgrad: Pack QKV gradients as well

### PR DESCRIPTION
**Pull request by danthe3rd**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* (to be filled)

### CODE REVIEW

File: xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu [hunk 1]
```diff
@@ -100,11 +100,25 @@ mem_efficient_attention_backward_cutlass(
   // keys with no query associated, so they are not
   // initialized
   bool grad_kv_needs_init = causal && N > M;
-  at::Tensor grad_q = at::empty_like(query);
-  at::Tensor grad_k =
-      grad_kv_needs_init ? at::zeros_like(key) : at::empty_like(key);
-  at::Tensor grad_v =
-      grad_kv_needs_init ? at::zeros_like(value) : at::empty_like(value);
+  at::Tensor grad_q, grad_k, grad_v;
+  if (!grad_kv_needs_init && query.size(1) == key.size(1) &&
```
Comment (fmassa) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu:104]: nit: looks like the `!grad_kv_needs_init` is redundant as `N > M` is always false if `query.size(1) == key.size(1)` ?

File: xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu [hunk 2]
```diff
@@ -100,11 +100,25 @@ mem_efficient_attention_backward_cutlass(
...
+      query.storage().is_alias_of(value.storage())) {
+    // Create one big contiguous chunk
+    // This is because q, k and v usually come from a single
+    // output of a linear layer that is chunked.
+    // Creating the gradients with the right layout saves us
+    // a `torch.cat` call in the backward pass
+    at::Tensor chunk = at::empty({B, M, 3, nH, K}, query.options());
+    grad_q = chunk.select(2, 0);
+    grad_k = chunk.select(2, 1);
+    grad_v = chunk.select(2, 2);
```
Comment (fmassa) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu:116]: Beautiful!

File: xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h [hunk 3]
```diff
@@ -83,9 +83,9 @@ struct AttentionBackwardKernel {
     accum_t* delta_ptr; // [Mq, nH]
 
     // Output tensors
-    scalar_t* grad_query_ptr; //  [Mq, nH, K]
-    scalar_t* grad_key_ptr; //    [Mk, nH, K]
-    scalar_t* grad_value_ptr; //  [Mk, nH, Kv]
+    output_t* grad_query_ptr; //  [Mq, nH, K]
+    output_t* grad_key_ptr; //    [Mk, nH, K]
+    output_t* grad_value_ptr; //  [Mk, nH, Kv]
```
Comment (fmassa) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h:88]: nit: I think this can be a bit misleading.

Indeed, if we define the `scalar_t` as the dtype of `query` / `key` / value`, and `output_t` the `dtype` of the output of the attention, then the gradients of `query` / `key` / `value` should have the same `dtype` as their respective `query` / `key` / `value`, and not on the output dtype.

If ever we make in the forward that `scalar_t != output_t`, then this could be an issue.

**Review by fmassa — APPROVED** [review]
This looks great, thanks Daniel!
